### PR TITLE
Add rowwise reductions to kernel generator

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Adds rowwise reductions (sum, min, max) to kernel generator.

## Tests
Rowwise reductions have their own set of tests.

## Side Effects
None.

## Checklist

- [x] Math issue #1342

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
Nice! I'll look this over more thoroughly tonight. On an initial pass, how does the local work group get setup here? Jw about like a 2,50 matrix that goes over the local work group size

**Comment by t4c1** [comment]
In all Stan Math, wherever no local communication is needed we don't specify local size. So it is decided by GPU driver. That includes whole kernel generator and this PR. Rowwise reductions do not use any local memory. Instead 1 thread calculates 1 result value.

All this will have to change a bit with colwise reductions, which will make a larger PR.

**Comment by syclik** [comment]
@t4c1: this has conflicts with `develop`.

**Comment by t4c1** [comment]
This is ready for review.

File: test/unit/math/opencl/kernel_generator/rowwise_reduction_test.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  MatrixXd m(3, 0);
+
+  matrix_cl<double> m_cl(m);
+
+  matrix_cl<double> res1_cl = stan::math::rowwise_sum(m_cl);
+  MatrixXd res1 = stan::math::from_matrix_cl(res1_cl);
+  MatrixXd correct1 = m.rowwise().sum();
+  EXPECT_EQ(correct1.rows(), res1.rows());
+  EXPECT_EQ(correct1.cols(), res1.cols());
+  EXPECT_MATRIX_NEAR(correct1, res1, 1e-9);
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/kernel_generator/rowwise_reduction_test.cpp]: I would like to see the following tests: 
- m(0,0)
- m(1,N)
- m(N,1)

File: stan/math/opencl/kernel_generator/operation_cl.hpp [hunk 2]
```diff
@@ -24,8 +24,10 @@ namespace math {
  * Parts of an OpenCL kernel, generated by an expression
  */
 struct kernel_parts {
-  std::string body;  // the body of the kernel - code executing operations
-  std::string args;  // kernel arguments
+  std::string body_start;  // the code that should be placed at the start of the
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/operation_cl.hpp]: [suggestion] Change name to body_prefix?

File: stan/math/opencl/kernel_generator/rowwise_reduction.hpp [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <string>
+#include <type_traits>
+#include <set>
+#include <utility>
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/rowwise_reduction.hpp]: Please order the headers alphabetically by first grouping them in prim, opencl, std headers groups.

**Pull request merged**